lib/balance-tx/lib/internal/Internal/Cardano/Write/Tx/TxWithUTxO/Gen.hs

```diff
@@ -0,0 +1,130 @@
+-- Copyright: © 2024 Cardano Foundation
+-- License: Apache-2.0
+--
+-- Provides generators and shrinkers for the 'TxWithUTxO' data type.
+--
+module Internal.Cardano.Write.Tx.TxWithUTxO.Gen
+    ( generate
+    , generateWithMinimalUTxO
+    , generateWithSurplusUTxO
+    , shrinkWith
+    , shrinkTxWith
+    , shrinkUTxOWith
+    )
+    where
+
+import Prelude
+
+import Cardano.Ledger.Api
+    ( EraTx (bodyTxL)
+    )
+import Cardano.Ledger.Api.Tx.Body
+    ( allInputsTxBodyF
+    )
+import Control.Lens
+    ( view
+    )
+import Internal.Cardano.Write.Tx
+    ( IsRecentEra
+    , Tx
+    , TxIn
+    , TxOut
+    , UTxO (UTxO)
+    )
+import Internal.Cardano.Write.Tx.TxWithUTxO
+    ( pattern TxWithUTxO
+    , type TxWithUTxO
+    )
+import Test.QuickCheck
+    ( Gen
+    , frequency
+    )
+import Test.QuickCheck.Extra
+    ( genMapFromKeysWith
+    , genNonEmptyDisjointMap
+    , interleaveRoundRobin
+    )
+
+import qualified Internal.Cardano.Write.Tx.TxWithUTxO as TxWithUTxO
+
+-- | Generates a 'TxWithUTxO' object.
+--
+-- The domain of the UTxO map is a superset of the transaction input set, but
+-- it may or may not be a strict superset.
+--
+generate
+    :: IsRecentEra era
+    => Gen (Tx era)
+    -> Gen (TxIn)
+    -> Gen (TxOut era)
+    -> Gen (TxWithUTxO era)
+generate genTx genTxIn genTxOut =
+    frequency
+        [ (9, generateWithMinimalUTxO genTx genTxIn genTxOut)
+        , (1, generateWithSurplusUTxO genTx genTxIn genTxOut)
+        ]
+
+-- | Generates a 'TxWithUTxO' object that has a minimal UTxO set.
+--
+-- The domain of the UTxO map is exactly equal to the transaction input set.
+--
+generateWithMinimalUTxO
+    :: IsRecentEra era
+    => Gen (Tx era)
+    -> Gen (TxIn)
+    -> Gen (TxOut era)
+    -> Gen (TxWithUTxO era)
+generateWithMinimalUTxO genTx _genTxIn genTxOut = do
+    tx <- genTx
+    utxo <- UTxO <$> genMapFromKeysWith genTxOut (txAllInputs tx)
+    pure $ TxWithUTxO.constructFiltered tx utxo
+  where
+    txAllInputs = view (bodyTxL . allInputsTxBodyF)
+
+-- | Generates a 'TxWithUTxO' object that has a surplus UTxO set.
+--
+-- The domain of the UTxO map is a strict superset of the transaction input set.
+--
+generateWithSurplusUTxO
+    :: forall era. ()
+    => IsRecentEra era
+    => Gen (Tx era)
+    -> Gen (TxIn)
+    -> Gen (TxOut era)
+    -> Gen (TxWithUTxO era)
+generateWithSurplusUTxO genTx genTxIn genTxOut =
+    generateWithMinimalUTxO genTx genTxIn genTxOut >>= \case
+        TxWithUTxO tx (UTxO utxo) -> do
+            utxoSurplus <- genNonEmptyDisjointMap genTxIn genTxOut utxo
+            pure $ TxWithUTxO.constructFiltered tx $ UTxO (utxo <> utxoSurplus)
+
+shrinkWith
+    :: IsRecentEra era
+    => (Tx         era -> [Tx         era])
+    -> (UTxO       era -> [UTxO       era])
+    -> (TxWithUTxO era -> [TxWithUTxO era])
+shrinkWith shrinkTx shrinkUTxO txWithUTxO =
+    interleaveRoundRobin
+        [ shrinkTxWith   shrinkTx   txWithUTxO
+        , shrinkUTxOWith shrinkUTxO txWithUTxO
```

Comment:
🤔 Looking at this I came to think: shouldn't an early shrinking step be dropping all surperflous UTxOs? And then: would there be a case to make for having `TxWithUTxO` _never_ contain surperflous UTxO entries? The only thing they affect would be the utxo inconsistency check.

Would depend on what your plans are, but perhaps food for thought for later.